[doc] Document how to contribute (#939)
apache/incubator-paimon: docs/content/project/contributing.md

---
title: "Contributing"
weight: 3
type: docs
aliases:
- /project/contributing.html
---
<!--
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

# Contributing

> Apache Paimon (incubating) is developed by an open and friendly community. Everybody is cordially welcome to join

Reviewer: ```suggestion
Apache Paimon (incubating) is developed by an open and friendly community. Everybody is cordially welcomed to join
```

Author: welcome is an adjective, here can be `welcome`

> the community and contribute to Apache Paimon. There are several ways to interact with the community and to contribute

Reviewer: ```suggestion
the community and contribute to Apache Paimon. There are several ways to interact with the community and contribute
```

> to Paimon including asking questions, filing bug reports, proposing new features, joining discussions on the mailing

Reviewer: ```suggestion
to Paimon including raising issues, filing bug reports, proposing new features, joining discussions on the mailing
```

Author: `raising issues` contains `filing bug reports`?

> lists, contributing code or documentation, improving the website, or testing release candidates.

Reviewer: ```suggestion
lists, contributing code or documentation, improving website, testing release candidates and writing corresponding blog etc.
```

    <div class="panel panel-default">
      <div class="panel-body">
        <h2><span class="number">3</span><a href="#review">Review</a></h2>
        <p>Work with the reviewer.</p><br />
        <p><b>1.Make sure no unrelated or unnecessary reformatting changes are included. 2.Make sure your change has been rebased to the latest commits in your base branch.</b></p>
      </div>
    </div>
  </div>
  <div class="column">
    <div class="panel panel-default">
      <div class="panel-body">
        <h2><span class="number">4</span><a href="#merge">Merge</a></h2>
        <p>A committer of Paimon checks if the contribution fulfills the requirements and merges the code to the codebase.</p>
      </div>
    </div>
  </div>
</div>

## Code Review Guide

Every review needs to check the following six aspects. **We encourage to check these aspects in order, to avoid
spending time on detailed code quality reviews when formal requirements are not met or there is no consensus in
the community to accept the change.**

#### 1. Is the Contribution Well-Described?

Check whether the contribution is sufficiently well-described to support a good review. Trivial changes and fixes

> do not need a long description. If the implementation is exactly according to a prior discussion on Issue or the

Reviewer: ```suggestion
do not need a long description. If the implementation is exactly according to a prior discussion on issue or the
```

development mailing list, only a short reference to that discussion is needed.

If the implementation is different from the agreed approach in the consensus discussion, a detailed description of
the implementation is required for any further review of the contribution.

#### 2. Does the Contribution Need Attention from some Specific Committers?

Some changes require attention and approval from specific committers.

If the pull request needs specific attention, one of the tagged committers/contributors should give the final approval.

#### 3. Is the Overall Code Quality Good, Meeting Standard we Want to Maintain in Paimon?

This is the detailed code review of the actual changes, covering:

> - Are the changes doing what is described in the Issue or design document?

Reviewer: ```suggestion
- Are the changes doing what is described in the issue or design document?
```

    flex-wrap: nowrap;
  }

  .contribute-grid .column {
    flex: 0 0 25%;
  }
}

.contribute-grid .panel {
  height: 100%;
  margin: 0;
}

.contribute-grid .panel-body {
  padding: 10px;
}

.contribute-grid h2 {
  margin: 0 0 10px 0;
  padding: 0;
  display: flex;
  align-items: flex-start;
}

.contribute-grid .number {
  margin-right: 0.25em;
  font-size: 1.5em;
  line-height: 0.9;
}
</style>

<div class="contribute-grid">
  <div class="column">
    <div class="panel panel-default">
      <div class="panel-body">
        <h2><span class="number">1</span><a href="#consensus">Discuss</a></h2>
        <p>Create a Issue or mailing list discussion and reach consensus</p>
        <p><b>To request an issue, please note that it is not just a "please assign it to me", you need to explain your understanding of the issue, and your design, and if possible, you need to provide your POC code.</b></p>
      </div>
    </div>
  </div>
  <div class="column">
    <div class="panel panel-default">
      <div class="panel-body">
        <h2><span class="number">2</span><a href="#implement">Implement</a></h2>
        <p>Create the Pull Request and the approach agreed upon in the issue.</p>

>         <p><b>1.Only create the PR if you are assigned to the issue. 2.Please associate an issue (if any), e.g. fix #123. 3.Please enable the actions of your own clone project, show a screenshot of the test passing.</b></p>

Reviewer 2: Why we must need to `show a screenshot of the test passing`?

Author: For non-apache-committers, only testing run in their own github repo. It is good to show the test passing.

Author: OK it is ugly, reviewer can take a look to corresponding link to see test passing.

      </div>
    </div>
  </div>
</div>

## Code Review Guide

Every review needs to check the following six aspects. **We encourage to check these aspects in order, to avoid
spending time on detailed code quality reviews when formal requirements are not met or there is no consensus in
the community to accept the change.**

#### 1. Is the Contribution Well-Described?

Check whether the contribution is sufficiently well-described to support a good review. Trivial changes and fixes
do not need a long description. If the implementation is exactly according to a prior discussion on Issue or the
development mailing list, only a short reference to that discussion is needed.

If the implementation is different from the agreed approach in the consensus discussion, a detailed description of
the implementation is required for any further review of the contribution.

#### 2. Does the Contribution Need Attention from some Specific Committers?

Some changes require attention and approval from specific committers.

If the pull request needs specific attention, one of the tagged committers/contributors should give the final approval.

> #### 3. Is the Overall Code Quality Good, Meeting Standard we Want to Maintain in Paimon?

Reviewer 2: I think the description of this part is a bit ambiguous. How can we define good code quality? I think we can add reference to [Flink's code guide](https://flink.apache.org/how-to-contribute/code-style-and-quality-preamble/) before Paimon already supports one.

Author: good point, add `Code guidelines can be found in the [Flink Code Style and Quality Guide](https://flink.apache.org/how-to-contribute/code-style-and-quality-preamble/).`
